### ISSUE TITLE
Duplicatable with associative tables (BelongsToMany Associations)

### ISSUE THREAD

**Issue by aritakku** [riesenia/cakephp-duplicatable#23]
Dear All,
I have a problem with associative table duplication.

Three related tables are (1) products, (2) offers, and (3) offer_rows. 

Offer rows have some extra data in addition of being the associative table between products and offers, and thus I had configured them previously as:

```
(class OffersTable)
@property \Cake\ORM\Association\HasMany $OfferRows

// initialize
$this->hasMany('OfferRows', [
    'foreignKey' => 'offer_id'
]);

(class ProductsTable)
@property \Cake\ORM\Association\BelongsToMany $OfferRows

// initialize
$this->belongsToMany('OfferRows', [
    'foreignKey' => 'product_id',
    'joinType' => 'INNER',
]);


(class OfferRowsTable)
@property \Cake\ORM\Association\BelongsTo $Offers
@property \Cake\ORM\Association\HasOne $Products

// initialize
$this->belongsTo('Offers', [
       'foreignKey' => 'offer_id',
       'joinType' => 'INNER'
]);        
$this->hasOne('Products', [
	'className' => 'Products',
        'propertyName' => 'reference_product_obj',
        'foreignKey' => 'reference_product'        		
]);


-- SQL

create table products(
	id serial not null auto_increment primary key,
	...
);
create table offers(
	id serial not null auto_increment primary key,
	...
);
create table offer_rows(
	id serial not null auto_increment primary key,

	product_id bigint(20) unsigned references products(id),
	offer_id bigint(20) unsigned not null references offers(id),

	...
);
```

With the abovementioned configuration the duplicatable worked perfectly with:

```        
$this->addBehavior('Duplicatable.Duplicatable', [
	'contain' => ['OfferRows'],
	'includeTranslations' => true,
	'remove' => ['created', 'modified', 'OfferRows.modified', 'OfferRows.created'],
	'set' => ['copied' => true],
 ]);
```

Then I modified the association description for cakephp to be:
(As to "why break working code", this modification was due to an error when deleting entity, which reflected some error in association configuration and I figured out that it should be fixed and thus changed it to BelongsToMany.)

```
(class OffersTable)
@property \Cake\ORM\Association\BelongsToMany $Products

// initialize
$this->belongsToMany('Products', [
     'through' => 'OfferRows',
     'foreignKey' => 'offer_id',
     'joinType' => 'INNER',
     'joinTable' => 'offer_rows',
]);


(class ProductsTable)
@property \Cake\ORM\Association\BelongsToMany $Offers

// initialize
$this->belongsToMany('Offers', [
     'through' => 'OfferRows',
     'foreignKey' => 'product_id',
     'joinType' => 'INNER',
     'joinTable' => 'offer_rows',
]);
        

(class OfferRowsTable)
@property \Cake\ORM\Association\BelongsTo $Products
@property \Cake\ORM\Association\BelongsTo $Offers

// initialize
$this->belongsTo('Products', [
     'foreignKey' => 'product_id'
]);
$this->belongsTo('Offers', [
     'foreignKey' => 'offer_id',
     'joinType' => 'INNER'
]);
```
Used versions are:
```
cakephp/cakephp                       3.3.7   The CakePHP framework
riesenia/cakephp-duplicatable         v1.0.5  CakePHP ORM plugin for duplicating entities (including related entities)
```

As a result, the offer duplication fails whenever I try to include the offer-to-products
information in the duplication. (Naturally I want to duplicate the offer rows with the offer, but not the products.)

I have tried to follow the instructions carefully from [cakephp documentation](https://book.cakephp.org/3.0/en/orm/associations.html#belongstomany-associations), but now I am wondering if this should work with duplicatable and how? 

(This new configuration has given also other headache - namely the fact that I cannot "link" the offer_rows to offer (with the "contain" => ["OfferRows"]), because one offer_row is accessible only through corresponding item in products-array (product."_joinData")).

**Comment by ADmad** [comment]
Did you try using latest release and setting contain to 'Products' in behavior config?

**Comment by aritakku** [comment]
I updated cakephp to version 3.3.16 and cakephp-duplicatable to v1.0.6.

```
OrdersTable.php:

$this->addBehavior('Duplicatable.Duplicatable', [
       		'contain' => ['Products'],
       		'includeTranslations' => true,
       		'remove' => ['created', 'modified',],
       		'set' => [ 'copied' => true	],
]);
```
Now after `$this->Offers->duplicate(<offer id>)` in (Controller) the offer is duplicated correctly and so are offer-to-product relations in the associate table. However, the other fields of the associated class is not. All (other fields than offer_id and product_id) in offer_rows are null. 

How do I configure which offer_rows fields I want to set to null and which to duplicate?

**Comment by ADmad** [comment]
> ... and cakephp-duplicatable to v1.0.6.

The latest release of this plugin is 3.0.5 :slightly_smiling_face:

**Comment by aritakku** [comment]
The problem persists with 3.0.5.

Also, I noted that a row which contains offer_id but not product_id is not duplicated at all.

**Comment by aritakku** [comment]
I am starting to suspect if I should use the `hasMany through` at all, because the offer_rows can contain row with null product_id. If there is no "end point" required for the associative table, shouldn't I use some other mapping than the `hasMany` ? :-o (I have also posted a related question to cakephp forum:  http://discourse.cakephp.org/t/blongstomany-configuration-hasmany-through-association/2779)

**Comment by ADmad** [comment]
You should first sort out your associations then we can look into whether there is actually a bug duplicatable behavior :)   I have used the behavior for `belongsToMany` association with `through` model without any issues.

If either of `product_id` or `offer_id` can be empty then it's not a many to many relation.

**Comment by aritakku** [comment]
Thanks for the help you have provided. 

I know what the DB looks like, but I am somewhat puzzled of how to map the DB to the cakephp. At first the OfferRows were just linked to the Offers, but then the relation to Products was added because that will enable quering the offered products. But because the product can be null, I am now asking for someone familiar with cakephp db modelling if I have error in my cakephp modelling.

**Comment by aritakku** [comment]
As it [turned out](https://stackoverflow.com/questions/44822157/cakephp-3-x-db-mapping-belongstomany-vs-hasmanyhasone), I had error in ORM mapping. Now the duplication works again!

However, after updating the cakephp files, redirect in the Controller after I have invoked the `duplicate()` went broke.

```
Warning (2): Invalid argument supplied for foreach() [ROOT/vendor/riesenia/cakephp-duplicatable/src/Model/Behavior/DuplicatableBehavior.php, line 269]
```
Code Context
```
Duplicatable\Model\Behavior\DuplicatableBehavior::_drillDownEntity() - ROOT/vendor/riesenia/cakephp-duplicatable/src/Model/Behavior/DuplicatableBehavior.php, line 269
Duplicatable\Model\Behavior\DuplicatableBehavior::duplicateEntity() - ROOT/vendor/riesenia/cakephp-duplicatable/src/Model/Behavior/DuplicatableBehavior.php, line 85
Duplicatable\Model\Behavior\DuplicatableBehavior::duplicate() - ROOT/vendor/riesenia/cakephp-duplicatable/src/Model/Behavior/DuplicatableBehavior.php, line 51
Cake\ORM\BehaviorRegistry::call() - CORE/src/ORM/BehaviorRegistry.php, line 237
Cake\ORM\Table::__call() - CORE/src/ORM/Table.php, line 2030
App\Model\Table\OffersTable::duplicate() - APP/Controller/OffersController.php, line 370 
```
I didn't find a reasonable cause for this, and so I ended up fixing this by editing the DuplicatableBehavior.php:
```
269:        if (!is_iterable($entity)) {
270:            return;
271:        }
272:
273:        foreach ($entity->{$prop} as $key => $e) {
274:             $this->_drillDownEntity($action, $e, $parts, $value);
275:        }
276:    }
```

(Function is_iterable is available [from PHP 7.1](http://php.net/manual/en/function.is-iterable.php), but for my purpose I use [this](https://stackoverflow.com/a/15655476/8230772) with the `function_exists` ).

**Comment by segy** [comment]
Hi. I am closing this after a while. If you think, there is a bug, please provide any test case. Thanks.
